src/lib/lib.jl

```diff
@@ -210,6 +210,42 @@ end
 @generated pair(::Val{k}, v, _=nothing) where k = :($k = v,)
 @generated pair(::Val{k}, v, ::NamedTuple{keys}) where {k,keys} = k isa Int ? :($(getfield(keys, k)) = v,) : :($k = v,)
 
+# ugly hack to make differentiating `getproperty` infer a lot better
+@generated function _pullback(cx::AContext, ::typeof(literal_getproperty), x, ::Val{f}) where f
+    sig(x) = Tuple{x, typeof(f)}
+    pb_sig(x) = Tuple{cx, typeof(getproperty), x, typeof(f)}
+
+    # either `getproperty` has a custom implementation or `_pullback(cx, ::getproperty, x, f)`
+    # is overloaded directly
+    is_getfield_fallback = which(_pullback, pb_sig(x)) == which(_pullback, pb_sig(Any)) && 
+        which(getproperty, sig(x)) == which(getproperty, sig(Any))
+
+    if is_getfield_fallback
+        # just copy pullback of `literal_getfield`
+        _sig = Tuple{cx, typeof(literal_getfield), x, Val{f}}
+        ci = code_lowered(_pullback, _sig)[1]
+
+        Meta.partially_inline!(
```

Comment:
Could you add a comment here explaining the intent behind this line?

Reply:
Ok, will do